### PR TITLE
Fix JWT realm chain fall through on missing request header 

### PR DESCRIPTION
**Pull request by albertzaharovits**
Other realms following a JWT realm in the realms chain should be
allowed to have their try at authenticating requests. Currently,
the first JWT realm in the chain will try to extract the credentials
from the request headers, and in case of failure, fail the request
altogether, denying subsequent realm's right to try to do their
authentication.

### CODE REVIEW

**Comment by albertzaharovits** [comment]
Here's how I've uncovered and tested it. I have used two realms the JWT one and a PKI one.
The problem shows up when the JWT comes first, before the PKI one, and the client uses TLS client authentication only (doesn't send the JWT HTTP headers). The request is rejected as such:
```
➜  elasticsearch-8.2.0-SNAPSHOT git:(jwt-realm-token) curl --cert-type PEM -k --cert ./config/certs/curl.pem --key ./config/certs/curl.key -i -X GET "https://localhost:9200/_security/_authenticate?pretty" -H 'Content-Type: application/json'
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Basic realm="security" charset="UTF-8"
WWW-Authenticate: Bearer realm="security"
WWW-Authenticate: ApiKey
content-type: application/json;charset=utf-8
content-length: 769

{
  "error" : {
    "root_cause" : [
      {
        "type" : "security_exception",
        "reason" : "error attempting to authenticate request",
        "header" : {
          "WWW-Authenticate" : [
            "Basic realm=\"security\" charset=\"UTF-8\"",
            "Bearer realm=\"security\"",
            "ApiKey"
          ]
        }
      }
    ],
    "type" : "security_exception",
    "reason" : "error attempting to authenticate request",
    "caused_by" : {
      "type" : "illegal_argument_exception",
      "reason" : "JWT bearer token must be non-null"
    },
    "header" : {
      "WWW-Authenticate" : [
        "Basic realm=\"security\" charset=\"UTF-8\"",
        "Bearer realm=\"security\"",
        "ApiKey"
      ]
    }
  },
  "status" : 401
}
```
On the flip side, if the PKI realm comes first, before the JWT one, and the client authenticates with the JWT headers and without client-side TLS authentication, the request is successfully authenticated, because the PKI realm yields when it encounters credentials it doesn't recognize.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java [hunk 1]
```diff
@@ -35,9 +35,8 @@
      * @param clientAuthenticationSharedSecret URL-safe Shared Secret for Client authentication. Required by some JWT realms.
      */
     public JwtAuthenticationToken(final SecureString endUserSignedJwt, @Nullable final SecureString clientAuthenticationSharedSecret) {
-        if (endUserSignedJwt == null) {
-            throw new IllegalArgumentException("JWT bearer token must be non-null");
-        } else if (endUserSignedJwt.isEmpty()) {
+        assert endUserSignedJwt != null;
```
Comment (justincr-elastic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java:38]: Does it make sense to revert this change? I am not sure if it is needed.

The null check in token() seems to be sufficient to fix the continuation bug going from JWT realm to non-JWT realm.
https://github.com/elastic/elasticsearch/blob/799a681906bfa35b8b94cd41b6a1e12625ff05d0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java#L291-L293

By changing `JwtAuthenticationToken` constructor, null JWT will still fail, but the descriptive error message is gone.
- Test: Null JWT triggers assert failure without the original descriptive error message.
- Main: Null JWT falls through and triggers an uncaught NullPointerException without the original descriptive error message on this line.
https://github.com/elastic/elasticsearch/blob/799a681906bfa35b8b94cd41b6a1e12625ff05d0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java#L49

**Review by justincr-elastic — APPROVED** [review]
The changes will fix the bug, but I think the JwtAuthenticationToken constructor change is not required and leads to less descriptive error messages. I will leave it to you to decide if you want to revert or clean up that part of the PR.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java (hunk 1 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java:38]: I can remove the assert of course. It's just a stylistic choice than anything else, I tend to use them.

**Pull request merged**